src/lib/Incubator/Cache/TagAware/AbstractTagAwareAdapter.php

```diff
@@ -0,0 +1,278 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+
+namespace EzSystems\SymfonyTools\Incubator\Cache\TagAware;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Log\LoggerAwareInterface;
+use Symfony\Component\Cache\Adapter\AdapterInterface;
+use Symfony\Component\Cache\Marshaller\DefaultMarshaller;
+use Symfony\Component\Cache\Marshaller\MarshallerInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+use Symfony\Component\Cache\ResettableInterface;
+use Symfony\Component\Cache\Traits\AbstractTrait;
+
+/**
+ * Re-implements Symfony's AbstractAdapter (as it uses private properties).
+ *
+ * In order to be able to store tags with values to avoid 2x lookups for tags, and to be able to use backported Marshaller
+ * for serialization.
+ */
+abstract class AbstractTagAwareAdapter implements AdapterInterface, LoggerAwareInterface, ResettableInterface
+{
+    use AbstractTrait;
+
+    protected const TAGS_PREFIX = "\0tags\0";
+
+    private $createCacheItem;
+    private $mergeByLifetime;
+    /**
+     * @var \Symfony\Component\Cache\Marshaller\MarshallerInterface
+     * NOTE: Not relevant in this way in Symfony 4+ where Abstract trait already uses this
+     */
+    protected $marshaller;
+
+    /**
+     * @var callable
+     *
+     * @todo Find a way to avoid this, a way that also works for FilesystemTagAwareAdapter.
+     *       It's currently like this in order to expose getId() from AbstractTrait to RedisTagAwareAdapter.
+     */
+    protected $getId;
+
+    /**
+     * @param string $namespace
+     * @param int $defaultLifetime
+     * @param MarshallerInterface|null $marshaller
+     *
+     * @throws \Symfony\Component\Cache\Exception\CacheException
+     */
+    protected function __construct($namespace = '', $defaultLifetime = 0, MarshallerInterface $marshaller = null)
+    {
+        $this->marshaller = $marshaller ?? new DefaultMarshaller();
+
+        $this->namespace = '' === $namespace ? '' : CacheItem::validateKey($namespace) . ':';
+        if (null !== $this->maxIdLength && \strlen($namespace) > $this->maxIdLength - 24) {
+            throw new InvalidArgumentException(sprintf('Namespace must be %d chars max, %d given ("%s")', $this->maxIdLength - 24, \strlen($namespace), $namespace));
+        }
+        $this->createCacheItem = \Closure::bind(
+            function ($key, $value, $isHit) use ($defaultLifetime) {
+                $item = new CacheItem();
+                $item->key = $key;
+                //<diff:AbstractAdapter> extract Value and Tags from the cache value
+                $item->value = $value['value'];
```

Comment:
Apparently, the index `value` may not exists if you change to this Adapter without clearing the cache before. 

```suggestion
                $item->value = $value['value'] ?? null;
```

Reply:
If you have not cleared cache it will not be an array and you will be missing info on tags.

So maybe we need to throw exception if `value` is missing saying you need to clear cache after enabling the adapter ;)